solana/solana-ibc/programs/solana-ibc/src/events.rs

```diff
@@ -5,42 +5,77 @@ use lib::hash::CryptoHash;
 /// Possible events emitted by the smart contract.
 ///
 /// The events are logged in their borsh-serialised form.
-#[derive(Clone, PartialEq, Eq, borsh::BorshSerialize, derive_more::From)]
+#[derive(
+    Clone,
+    PartialEq,
+    Eq,
+    borsh::BorshSerialize,
+    borsh::BorshDeserialize,
+    derive_more::From,
+)]
 pub enum Event<'a> {
     IbcEvent(ibc::core::events::IbcEvent),
     Initialised(Initialised<'a>),
     NewBlock(NewBlock<'a>),
-    BlockSigned(BlockSigned<'a>),
+    BlockSigned(BlockSigned),
     BlockFinalised(BlockFinalised<'a>),
 }
 
 /// Event emitted once blockchain is implemented.
-#[derive(Clone, PartialEq, Eq, borsh::BorshSerialize, derive_more::From)]
+#[derive(
+    Clone,
+    PartialEq,
+    Eq,
+    borsh::BorshSerialize,
+    borsh::BorshDeserialize,
+    derive_more::From,
+)]
 pub struct Initialised<'a> {
     /// Genesis block of the chain.
     pub genesis: NewBlock<'a>,
 }
 
 /// Event emitted once a new block is generated.
-#[derive(Clone, PartialEq, Eq, borsh::BorshSerialize, derive_more::From)]
+#[derive(
+    Clone,
+    PartialEq,
+    Eq,
+    borsh::BorshSerialize,
+    borsh::BorshDeserialize,
+    derive_more::From,
+)]
 pub struct NewBlock<'a> {
     /// Hash of the new block.
-    pub hash: &'a CryptoHash,
+    pub hash: CryptoHash,
     /// The new block.
-    pub block: &'a crate::chain::Block,
+    pub block: CowBlock<'a>,
 }
 
 /// Event emitted once a new block is generated.
-#[derive(Clone, PartialEq, Eq, borsh::BorshSerialize, derive_more::From)]
-pub struct BlockSigned<'a> {
+#[derive(
+    Clone,
+    PartialEq,
+    Eq,
+    borsh::BorshSerialize,
+    borsh::BorshDeserialize,
+    derive_more::From,
+)]
+pub struct BlockSigned {
     /// Hash of the block to which signature was added.
-    pub block_hash: &'a CryptoHash,
+    pub block_hash: CryptoHash,
     /// Public key of the validator whose signature was added.
-    pub pubkey: &'a crate::chain::PubKey,
+    pub pubkey: crate::chain::PubKey,
 }
 
 /// Event emitted once a block is finalised.
-#[derive(Clone, PartialEq, Eq, borsh::BorshSerialize, derive_more::From)]
+#[derive(
+    Clone,
+    PartialEq,
+    Eq,
+    borsh::BorshSerialize,
+    borsh::BorshDeserialize,
+    derive_more::From,
+)]
 pub struct BlockFinalised<'a> {
     /// Hash of the block to which signature was added.
     pub block_hash: &'a CryptoHash,
```

Comment:
We still have the reference to CryptoHash here. I think we should remove the reference here as well.
```suggestion
pub struct BlockFinalised {
    /// Hash of the block to which signature was added.
    pub block_hash:  CryptoHash,
```

Reply:
And also clone the hash value so that we dont take the reference here.
https://github.com/ComposableFi/emulated-light-client/blob/9a01809be249e231804f568cd67200d59548cc9d/solana/solana-ibc/programs/solana-ibc/src/chain.rs#L122